enhancements/multi-arch/heterogeneous-architecture.md

```diff
@@ -0,0 +1,337 @@
+The trigger for creating the heterogeneous payloads will be whenever there is a new payload produced for one of the included architectures.
+
+In the case of `oc adm release new`, the manifestlist flow would be triggered whenever the CVO image in a source image/imagestream is a manifestlist.
+- If the source CVO image is a standard manifest, the generated release payload will also be a manifest.
+- If the source CVO image is a manifestlist, the generated release payload would be a manifestlist(containing a manifest for each architecture
+possessed by the CVO manifestlist).
+
+In either case, oc adm release new would permit non-CVO component images to be manifest or manifestlists and pass them through directly to the
+resultant release manifest(s). If a manifestlist release payload is generated, each architecture specific release payload manifest will reference
+the same pullspecs provided in the input imagestream.
+
+Since the internal Openshift registry does not support manifestlists, the release controller would use quay.io. There will be a new architecture option
+added to Cincinnati called "multi".
+
+### Upgrades - Cincinnati and CVO
+
+In addition to the homogeneous payload graph that we have today, [support][OCPPLAN-7642] needs to be added for an additional graph for heterogeneous
+payloads which CVO will need to look at to make upgrade recommendations. There are a few upgrade scenarios to consider with the addition of
+heterogeneous payloads:
+
+- heterogeneous -> heterogeneous payload upgrade
+- homogeneous -> heterogeneous payload upgrade
+- heterogeneous -> homogeneous payload upgrade
+
+The support for homogeneous payloads will exist so customers running disconnected homogeneous clusters do not incur the cost of mirroring the
+additional architectures if they were to use heterogeneous payloads. This means that in order for CVO to find the current release in the graph,it
+will have to know about the existing homogeneous graph as well in addition to the new heterogeneous graph. There are few options on the best way
+to make update recommendations especially in the case of upgrading from a homogeneous to a heterogeneous payload:
+
+- Have the heterogeneous graph include the homogeneous releases so CVO can make recommendations based on the current homogeneous release image.
+This is complicated by the fact that there will be multiple releases with the same version but different payloads. The trigger to update to a
+heterogeneous payload would be an option provided by the administrator through either the ClusterVersion or Infrastructure API to "move" to the
+heterogeneous graph.
+- Have both the graphs be completely independent of each other and have an option in one of the aforementioned APIs to make the cluster heterogeneous
+and CVO would request the heterogeneous graph, find the current release version and update to it.
```

Comment:
another option would be to have an entirely separate channel for heterogeneous payloads (e.g. `stable-multiarch-4.11`) and make admins switch channels if they want to move between heterogeneous and homogeneous payloads.

Reply:
i added this. so i guess the trigger for switching channels would also be the new field we introduce in the ClusterVersion or Infra resource to say "move to a heterogeneous" payload? or would admins have to specify the channel as well?

Reply:
my idea was the admins would have to specify/change the channel explicitly.  not much point in having a new api field that modifies the subscribed channel.

Reply:
i'd expect @sdodson and @wking to have the strongest opinions on whether a new multi-arch channel is appropriate/better/worse than the current api field proposal.

Reply:
Currently channel selection encodes:

* A cap minor version, e.g. "I want to hear about updates to 4.9 and earlier, but not about updated into 4.10 and later".
* A hint at risk appetite for update-edge and target-release stability (`candidate`, `fast`, `stable`, and `eus`).

I'm not excited about mixing in architecture decisions there, although I agree that overloading channel like that might help avoid growing a new knob for heterogeneous vs. homogeneous.  With the channel approach, the transition would look like:

1. User is running version A on homogeneous `stable-4.11`.
2. User decides to move to heterogeneous releases.
3. User sets a `stable-multiarch-4.11` channel.
4. CVO fetches that graph.  Today we'd successfully find the current release and avoid `VersionNotFound` because the CVO [naively assumes that it can match on the version string, without worrying if the image digest matches][1].
5. User has to update within this graph to version B, or explicitly `oc adm upgrade --allow-explicit-upgrade --to-image "${BY_DIGEST_PULLSPEC_FOR_HETEROGENOUS_RELEASE_A}"` to make the transition.

With the arch approach, the flow is:

1. User is running version A on homogeneous `stable-4.11`.
2. User decides to move to heterogeneous releases.
3. User sets the property in Infrastructure or ClusterVersion or wherever.
4. CVO fetches that graph, finds the current release A with the new architecture, and initiates an "update" to roll it out.
5. If the user wants an update to release B as well, they can kick that off now completely within the heterogeneous graph.

[1]: https://github.com/openshift/cluster-version-operator/blob/3db74ebb012838ffdd776ca65210e4d6d66b7472/pkg/cincinnati/cincinnati.go#L127-L146

Reply:
The decision to move to a multiarch should a configuration change for the user. The UX of only changing a channel does not fit in that model. I think this flow would work better from the UX point of view
```
With the arch approach, the flow is:

User is running version A on homogeneous stable-4.11.
User decides to move to heterogeneous releases.
User sets the property in Infrastructure or ClusterVersion or wherever.
CVO fetches that graph, finds the current release A with the new architecture, and initiates an "update" to roll it out.
If the user wants an update to release B as well, they can kick that off now completely within the heterogeneous graph
```